File: mobile/flutter_app/lib/features/exercises/domain/models/plan_modification.dart
class PlanModification {
  final int id;
  final int userId;
  final int? planId;
  final String modificationType;
  final String targetType;
  final String? targetDayLabel;
  final String? targetBlockTitle;
  final String? targetItemName;
  final String? replacementItemName;
  final String overrideJson;
  final String? noteText;
  final bool isActive;
  final DateTime? createdAt;

  const PlanModification({
    required this.id,
    required this.userId,
    this.planId,
    required this.modificationType,
    required this.targetType,
    this.targetDayLabel,
    this.targetBlockTitle,
    this.targetItemName,
    this.replacementItemName,
    required this.overrideJson,
    this.noteText,
    required this.isActive,
    this.createdAt,
  });

  factory PlanModification.fromJson(Map<String, dynamic> json) {
    return PlanModification(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      planId: json['plan_id'] as int?,
      modificationType: json['modification_type'] as String? ?? '',
      targetType: json['target_type'] as String? ?? '',
      targetDayLabel: json['target_day_label'] as String?,
      targetBlockTitle: json['target_block_title'] as String?,
      targetItemName: json['target_item_name'] as String?,
      replacementItemName: json['replacement_item_name'] as String?,
      overrideJson: json['override_json'] as String? ?? '{}',
      noteText: json['note_text'] as String?,
      isActive: json['is_active'] as bool? ?? true,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'modification_type': modificationType,
        'target_type': targetType,
        if (targetDayLabel != null) 'target_day_label': targetDayLabel,
        if (targetBlockTitle != null) 'target_block_title': targetBlockTitle,
        'target_item_name': targetItemName,
        if (replacementItemName != null)
          'replacement_item_name': replacementItemName,
        'override_json': overrideJson,
        if (noteText != null) 'note_text': noteText,
      };
}

File: backend/api/app/models/plan_modification.py
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class PlanModification(Base):
    __tablename__ = "plan_modifications"
    __table_args__ = (
        Index("ix_plan_modifications_user_id", "user_id"),
        Index("ix_plan_modifications_is_active", "is_active"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    plan_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        ForeignKey("initial_plans.id", ondelete="SET NULL"),
        nullable=True,
    )
    modification_type: Mapped[str] = mapped_column(String(32), nullable=False)
    # exclude / swap / edit_params / add_note
    target_type: Mapped[str] = mapped_column(String(32), default="exercise", nullable=False)
    # exercise / meal
    target_day_label: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    target_block_title: Mapped[Optional[str]] = mapped_column(String(120), nullable=True)
    target_item_name: Mapped[Optional[str]] = mapped_column(String(240), nullable=True)
    replacement_item_name: Mapped[Optional[str]] = mapped_column(String(240), nullable=True)
    override_json: Mapped[str] = mapped_column(Text, default="{}", nullable=False)
    note_text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
    )
